pkg/controller/controllername/controllername.go

```diff
@@ -3,37 +3,49 @@ package controllername
 import (
 	"strings"
 	"unicode"
+
+	"github.com/go-logr/logr"
 )
 
 const (
 	metricsNameDelimiter = "_"
 	loggerNameDelimiter  = "-"
 )
 
+// ControllerNamer is an interface that returns the name of the controller in all necessary forms
 type ControllerNamer interface {
+	// String returns the name of the controller in a human-readable form
+	String() string
+	// MetricsName returns the name of the controller in a form that can be used for Prometheus metrics, specifically as a Prometheus label https://prometheus.io/docs/practices/naming/#labels
 	MetricsName() string
+	// LoggerName returns the name of the controller in a form that can be used for logr logger naming
 	LoggerName() string
+	// AddToLogger adds controller name fields to the logger then returns the logger with the added fields
+	AddToLogger(l logr.Logger) logr.Logger
 }
 
 // controllerName ex. {"My","Controller", "Name"} ->  MyControllerName
 type controllerName []string
 
-func NewControllerName(name []string) controllerName {
-	cn := make(controllerName, len(name))
-
-	for i, w := range name {
-		cn[i] = strip(strings.ToLower(w))
-
+// New returns a new controllerName after taking each word of the controller name as a separate argument
+func New(firstWord string, additionalWords ...string) controllerName { // using a non-variadic for the first word makes it impossible to accidentally pass no arguments in. Accepting variadic versus slices also helps with not accepting empty slices and is easier to use
```

Comment:
what happens if the first word is an empty string and there's no variadic inputs?

Reply:
then the controller name is just an empty string which isn't that bad. The metrics label for the controller will be "" and the logger fields will be "" which doesn't break anything. 

We try to avoid that as much as possible with the entire args not being variadic which means someone has to call this like
```Go
controllername.New("")

```
which should obviously look wrong to any user and should be caught during any code review hopefully

Reply:
I think that would result in a blank name but feel like that's a pretty niche use-case? I guess we could add an error to the return and error if it's blank?

Reply:
this is as good as it's going to get without a linter (overkill for this) or runtime verification (can't have because we shouldn't crash operator due to a bad name). We can't fully protect against a consumer consuming this package completely wrong but that's pretty much true of every single package, there's only so much you can do.

Reply:
sounds good. i agree this provides the most protection